### PR TITLE
[octavia-ingress-controller] support for whitelist-source-range annotation

### LINKED ISSUE #1703: [octavia-ingress-controller] support whitelist-source-range annotation
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind feature

**What happened**:

I want to be able to use the `octavia.ingress.kubernetes.io/whitelist-source-range` annotation.

By default, Load Balancers are open to the world (`0.0.0.0/0`). If someone wants to restrict access to certain IP addresses, one should be able to use Octavia's `allowed_cidrs` feature through this Ingress annotation.

### PR DESCRIPTION
**Pull request by hedinasr**
**What this PR does / why we need it**:
Add support for `ingress.kubernetes.io/whitelist-source-range` annotation.
**Which issue this PR fixes(if applicable)**:
fixes #1703 

**Special notes for reviewers**:

**Release note**:

```release-note
[octavia-ingress-controller] Add support for whitelist-source-range annotation.
```

### CODE REVIEW

File: pkg/ingress/controller/controller.go [hunk 1]
```diff
@@ -95,6 +95,10 @@ const (
 	// Default to true.
 	IngressAnnotationInternal = "octavia.ingress.kubernetes.io/internal"
 
+	// IngressAnnotationSourceRangesKey is the key of the annotation on an ingress to set allowed IP ranges on their LoadBalancers.
+	// It should be a comma-separated list of CIDRs.
+	IngressAnnotationSourceRangesKey = "ingress.kubernetes.io/whitelist-source-range"
```
Comment (zetaab) [inline pkg/ingress/controller/controller.go:100]: maybe better name for this is `octavia.ingress.kubernetes.io/whitelist-source-range`? because its octavia ingress controller specific setting, not whole kubernetes

Comment (hedinasr) [reply pkg/ingress/controller/controller.go:100]: you're totally right, as others annotations are also prefixed by `octavia.ingress.kubernetes.io`.

**Comment by ramineni** [comment]
@hedinasr please update the release note as well .

**Comment by hedinasr** [comment]
Any news on this PR please ?

File: docs/octavia-ingress-controller/using-octavia-ingress-controller.md [hunk 2]
```diff
@@ -449,3 +450,33 @@ Ingress and enable the more secure HTTPS protocol.
...
 > in the future.
+
+## Allow CIDRs
+
+You can use the key `octavia.ingress.kubernetes.io/whitelist-source-range` annotations to whitelist CIDRs.
+The value should be a comma-separated list of CIDRs.
+
+Example:
+
+    ```yaml
```
Comment (lingxiankong) [inline docs/octavia-ingress-controller/using-octavia-ingress-controller.md:461]: Please remove the indentation of this yaml content.

File: docs/octavia-ingress-controller/using-octavia-ingress-controller.md [hunk 3]
```diff
@@ -449,3 +450,33 @@ Ingress and enable the more secure HTTPS protocol.
 > NOTE: octavia-ingress-controller currently doesn't support to integrate with
 > `cert-manager` to create the non-existing secret dynamically. Could be improved
 > in the future.
+
+## Allow CIDRs
+
+You can use the key `octavia.ingress.kubernetes.io/whitelist-source-range` annotations to whitelist CIDRs.
```
Comment (lingxiankong) [inline docs/octavia-ingress-controller/using-octavia-ingress-controller.md:456]: Personally, I think the description in the issue page looks more user friendly than here.

File: pkg/ingress/controller/controller.go [hunk 4]
```diff
@@ -695,7 +703,12 @@ func (c *Controller) ensureIngress(ing *nwv1.Ingress) error {
 	}
 
 	// Create listener
-	listener, err := c.osClient.EnsureListener(resName, lb.ID, secretRefs)
+	sourceRanges := getStringFromIngressAnnotation(ing, IngressAnnotationSourceRangesKey, "0.0.0.0/0")
+	listenerAllowedCIDRs := strings.Split(sourceRanges, ",")
+	if err != nil {
```
Comment (lingxiankong) [inline pkg/ingress/controller/controller.go:708]: I didn't see there is any err raised above.

File: pkg/ingress/controller/openstack/octavia.go [hunk 5]
```diff
@@ -356,12 +356,26 @@ func (os *OpenStack) EnsureListener(name string, lbID string, secretRefs []strin
...
+			opts.AllowedCIDRs = listenerAllowedCIDRs
+		}
 		listener, err = listeners.Create(os.Octavia, opts).Extract()
 		if err != nil {
 			return nil, fmt.Errorf("error creating listener: %v", err)
 		}
 
 		log.WithFields(log.Fields{"lbID": lbID, "listenerName": name}).Info("listener created")
+	} else {
+		if len(listenerAllowedCIDRs) > 0 {
```
Comment (lingxiankong) [inline pkg/ingress/controller/openstack/octavia.go:369]: What if the allowed_cidrs is not changed?

**Review by lingxiankong — CHANGES_REQUESTED** [review]

File: pkg/ingress/controller/controller.go [hunk 6]
```diff
@@ -312,6 +316,10 @@ func NewController(conf config.Config) *Controller {
 				// Two different versions of the same Ingress will always have different RVs.
 				return
 			}
+			newAnnotations := newIng.ObjectMeta.Annotations
+			oldAnnotations := oldIng.ObjectMeta.Annotations
+			delete(newAnnotations, "kubectl.kubernetes.io/last-applied-configuration")
+			delete(oldAnnotations, "kubectl.kubernetes.io/last-applied-configuration")
```
Comment (jichenjc) [inline pkg/ingress/controller/controller.go:322]: what's the purpose for this? seems not related to the issue itself?

Comment (hedinasr) [reply pkg/ingress/controller/controller.go:322]: Without this, a changement made in the annotations will not be applied.

File: docs/octavia-ingress-controller/using-octavia-ingress-controller.md (hunk 2 above)
Comment (hedinasr) [reply docs/octavia-ingress-controller/using-octavia-ingress-controller.md]: Fixed by https://github.com/kubernetes/cloud-provider-openstack/pull/1704/commits/06acf15e30d6bd467d32ff13c4fbbd30ecc9b439.

File: docs/octavia-ingress-controller/using-octavia-ingress-controller.md (hunk 3 above)
Comment (hedinasr) [reply docs/octavia-ingress-controller/using-octavia-ingress-controller.md]: Fixed by https://github.com/kubernetes/cloud-provider-openstack/pull/1704/commits/06acf15e30d6bd467d32ff13c4fbbd30ecc9b439.

File: pkg/ingress/controller/controller.go (hunk 4 above)
Comment (hedinasr) [reply pkg/ingress/controller/controller.go]: Fixed by https://github.com/kubernetes/cloud-provider-openstack/pull/1704/commits/06acf15e30d6bd467d32ff13c4fbbd30ecc9b439.

File: pkg/ingress/controller/openstack/octavia.go (hunk 5 above)
Comment (hedinasr) [reply pkg/ingress/controller/openstack/octavia.go:369]: You mean that this will occur whether the annotation is updated or not ?

File: pkg/ingress/controller/controller.go (hunk 6 above)
Comment (jichenjc) [reply pkg/ingress/controller/controller.go:322]: then I am curious why this is the first time we introduce this given we have so many annotations already

File: pkg/ingress/controller/openstack/octavia.go (hunk 5 above)
Comment (jichenjc) [reply pkg/ingress/controller/openstack/octavia.go:369]: I guess if the listenerAllowedCIDRs is not changed in this reconcile cycle
no need to execute any command below as it's kept?

File: pkg/ingress/controller/controller.go (hunk 6 above)
Comment (hedinasr) [reply pkg/ingress/controller/controller.go:322]: from what I'v seen, the `UpdateEvent` was not triggered when an annotation were updated. Hence these lines, and the condition line 333.

```golang
!reflect.DeepEqual(newAnnotations, oldAnnotations)
```

File: pkg/ingress/controller/openstack/octavia.go (hunk 5 above)
Comment (hedinasr) [reply pkg/ingress/controller/openstack/octavia.go]: should be fixed by https://github.com/kubernetes/cloud-provider-openstack/pull/1704/commits/75f28b380004ee3dba697521d87f92bde4f128e0.

File: pkg/ingress/controller/controller.go (hunk 6 above)
Comment (jichenjc) [reply pkg/ingress/controller/controller.go:322]: ok, thanks for the info~
